src/StoreApi/Routes/Checkout.php

```diff
@@ -320,6 +333,121 @@ protected function create_or_update_draft_order() {
 		return $order_object;
 	}
 
+	/**
+	 * Check request options and store (shop) config to determine if a user account
+	 * should be created as part of order processing.
+	 *
+	 * @param \WP_REST_Request $request The current request object being handled.
+	 *
+	 * @return boolean True if a new user account should be created.
+	 */
+	protected function should_create_customer_account( $request ) {
+		if ( is_user_logged_in() ) {
+			return false;
+		}
+
+		$checkout_requires_account = false === filter_var( get_option( 'woocommerce_enable_guest_checkout' ), FILTER_VALIDATE_BOOLEAN );
+		if ( $checkout_requires_account ) {
+			return true;
+		}
+
+		$user_requested_account = ! empty( $request['create_account'] ) && true === filter_var( $request['create_account'], FILTER_VALIDATE_BOOLEAN );
+		if ( $user_requested_account ) {
+			return true;
+		}
+
+		return false;
+	}
+
+	/**
+	 * Convert a account creation error to a Store API error.
+	 *
+	 * @param \WP_Error $error An error object.
+	 *
+	 * @return RouteException API error object with error details.
+	 */
+	private function map_create_account_error( \WP_Error $error ) {
+		switch ( $error->get_error_code() ) {
+			// WordPress core error codes.
+			case 'empty_username':
+			case 'invalid_username':
+			case 'empty_email':
+			case 'invalid_email':
+			case 'email_exists':
+			case 'registerfail':
+				return new RouteException(
+					'woocommerce_rest_checkout_create_account_failure',
+					wp_strip_all_tags( $error->get_error_message() ),
```

Comment:
We're stripping all HTML so any error messages are plain text, for our notice API. 

This was originally added because [WooCommerce Core generates errors with log in links](https://github.com/woocommerce/woocommerce/blob/e736ac4145d9aceb3574922d01824ede546a07f3/includes/wc-user-functions.php#L47) and we don't have a way to render those safely in blocks `Notice` system (yet). 

This strip might not be needed (I haven't checked all the core errors), but it's safer to leave it in place.

Regarding the `Please log in.` link - this would be a nice improvement to a specific error case, for example:

- Shopper is already signed up, but not logged in. (Maybe they forgot they had an account.)
- Shopper completes a purchase and requests an account via checkout.
- Shopper sees error notice "You already have an account - please log in".

In Gutenberg [`Notice` there is support for an `notice actions`](https://github.com/WordPress/gutenberg/tree/master/packages/components/src/notice), which we could use to add `Log in` CTA/button/link to notices. We'd need to support it in our Notice API, so I've left this as follow up for now - keen for feedback. (I can make an issue if we think this is worth pursuing cc @nerrad.)

Reply:
https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/2985